File: Selection_Sort/R/ducthanhtran/selection_sort.R
# Selection Sort algorithm
# Author: Duc Thanh Tran (ducthanhtran, Github)
# Date: 01/26/2015

# selection sort algorithm
selectionSort <- function(arr) {
  for (i in 1:(length(arr)-1)) {
    indexMin <- i;
    
    for (j in (i+1):length(arr)) {
      if(arr[j] < arr[indexMin])
        indexMin <- j;
    }
    
    if(indexMin != i)
      arr <- swap(arr, indexMin, i);
  }
  return(arr);
}


# swap function
# swaps values at indices indexA and indexB
swap <- function(arr, indexA, indexB) {
  temp <- arr[indexA];
  arr[indexA] <- arr[indexB];
  arr[indexB] <- temp;
  return(arr);
}

# test algorithm
arrA <- c(9, 2, 4, 5, 1);
print(arrA);

arrResult <- selectionSort(arrA);
print(arrResult);

File: Selection_Sort/Lua/ducthanhtran/selectionsort.lua
--[[
	Author: Duc Thanh Tran (ducthanhtran, Github)
	Algorithm: Selectionsort, basic version where minimum element is searched for,
               not stable version!
	Date: 12/31/2013
--]]

local function selectionsort(T)
	assert(type(T) == "table", "Argument is not a table")

	local length = #T

	for i = 1, length-1 do
		local minElem = i

		for j = i+1, length do
			if T[j] < T[minElem] then
				minElem = j
			end
		end

		T[i], T[minElem] = T[minElem], T[i]
	end
end

return selectionsort
